### PR TITLE
[Pure k8s] add scheduler validation in restserver

### PR DESCRIPTION
**Pull request by mzmssg**
1. Add deployment flag to enable/disable hived
2. Validate memory/cpu in hived mode
3. Add resource unit api to query cluster sku

### CODE REVIEW

File: src/rest-server/src/config/v2/hived.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+let enabledHived = false;
+if (process.env.LAUNCHER_TYPE === 'k8s' && process.env.SCHEDULER_TYPE === 'hived') {
+  enabledHived = true;
+}
+
+let hivedSpecPath = '/hived-spec/hivedscheduler.yaml';
+if (process.env.HIVED_SPEC_PATH) {
+  hivedSpecPath = process.env.HIVED_SPEC_PATH;
```
Comment (abuccts) [inline src/rest-server/src/config/v2/hived.js]: where will you put this file?

File: src/rest-server/src/config/vc.js [hunk 2]
```diff
@@ -37,8 +40,47 @@ const vcStatusPutInputSchema = Joi.object().keys({
...
+      memoryMb /= 1000;
+      break;
+    default:
+      memoryMb /= 1000000;
+  }
+  return memoryMb;
+};
+
+for (let gpuType of Object.keys(resourceUnits)) {
+  resourceUnits[gpuType].memoryMB = convertMemory(resourceUnits[gpuType].memory);
```
Comment (abuccts) [inline src/rest-server/src/config/vc.js]: why cannot use `memoryMB` directly?

File: src/rest-server/src/models/v2/job/k8s.js [hunk 3]
```diff
@@ -310,6 +292,33 @@ const generateTaskRole = (taskRole, labels, config) => {
...
+  if (hivedConfig.enabledHived) {
+    frameworkTaskRole.task.pod.spec.schedulerName = launcherConfig.scheduler;
+
+    delete frameworkTaskRole.task.pod.spec.containers[0].resources.limits['nvidia.com/gpu'];
+    frameworkTaskRole.task.pod.spec.containers[0]
+      .resources.limits['hivedscheduler.microsoft.com/pod-scheduling-enable'] = 1;
+
+    if (frameworkTaskRole.task.pod.metadata.annotations == null) {
+      frameworkTaskRole.task.pod.metadata.annotations = {};
+    }
```
Comment (abuccts) [inline src/rest-server/src/models/v2/job/k8s.js]: unnecessary

File: src/rest-server/src/models/v2/job/k8s.js [hunk 4]
```diff
@@ -310,6 +292,33 @@ const generateTaskRole = (taskRole, labels, config) => {
...
+    frameworkTaskRole.task.pod.metadata.annotations['hivedscheduler.microsoft.com/pod-scheduling-spec'] = yaml.safeDump(config.taskRoles[taskRole].hivedPodSpec);
+
+    if (frameworkTaskRole.task.pod.spec.containers[0].env == null) {
+      frameworkTaskRole.task.pod.spec.containers[0].env = [];
+    }
```
Comment (abuccts) [inline src/rest-server/src/models/v2/job/k8s.js]: same

File: src/rest-server/src/models/v2/virtual-cluster/k8s.js [hunk 5]
```diff
@@ -15,3 +15,23 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
+// module dependencies
+const vcConfig = require('@pai/config/vc');
+
+
+class VirtualCluster {
```
Comment (abuccts) [inline src/rest-server/src/models/v2/virtual-cluster/k8s.js]: please use functions directly

File: src/rest-server/src/models/v2/virtual-cluster/index.js [hunk 6]
```diff
@@ -20,5 +20,10 @@ const launcherConfig = require('@pai/config/launcher');
 if (launcherConfig.type === 'yarn') {
   module.exports = require('@pai/models/v2/virtual-cluster/yarn');
 } else {
-  module.exports = require('@pai/models/v2/virtual-cluster/yarn');
+  const yarnModel = require('@pai/models/v2/virtual-cluster/yarn');
```
Comment (abuccts) [inline src/rest-server/src/models/v2/virtual-cluster/index.js]: remove

File: src/rest-server/src/models/v2/virtual-cluster/yarn.js [hunk 7]
```diff
@@ -228,6 +229,11 @@ class VirtualCluster {
     });
   }
 
+  getResourceUnits() {
+    // TODO: get it from yarn or cluster configuration
+    return vcConfig.resourceUnits;
```
Comment (abuccts) [inline src/rest-server/src/models/v2/virtual-cluster/yarn.js]: throw error here

File: src/rest-server/src/config/v2/hived.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+
+// module dependencies
+const Ajv = require('ajv');
+const ajv = new Ajv({allErrors: true, useDefaults: true});
+
+
+let enabledHived = false;
+if (process.env.LAUNCHER_TYPE === 'k8s' && process.env.SCHEDULER_TYPE === 'hived') {
+  enabledHived = true;
```
Comment (abuccts) [inline src/rest-server/src/config/v2/hived.js]: is this related to job or vc?

File: src/rest-server/src/middlewares/v2/protocol.js [hunk 9]
```diff
@@ -70,6 +72,12 @@ const convertPriority = (priorityClass) => {
...
 const hivedValidate = (protocolObj) => {
+  if (!hivedSchema.validate(protocolObj)) {
+    throw createError('Bad Request', 'InvalidProtocolError', hivedSchema.validate.errors);
+  }
+  const resourceUnits = vcConfig.resourceUnits;
+  const minCpu = Math.min(...Array.from(Object.values(resourceUnits), (v)=>v.cpu));
+  const minMemoryMB = Math.min(...Array.from(Object.values(resourceUnits), (v)=>v.memoryMB));
   let hivedConfig = null;
   const affinityGroups = {};
   if ('extras' in protocolObj && 'hivedScheduler' in protocolObj.extras) {
```
Comment (abuccts) [inline src/rest-server/src/middlewares/v2/protocol.js]: remove `console`

File: src/rest-server/src/models/v2/virtual-cluster/k8s.js [hunk 10]
```diff
@@ -15,3 +15,23 @@
...
+  getResourceUnits() {
+    return this.resourceUnits;
+  }
+}
+
+const vc = new VirtualCluster();
+
+// module exports
+module.exports = {
+  getResourceUnits: vc.getResourceUnits.bind(vc),
```
Comment (abuccts) [inline src/rest-server/src/models/v2/virtual-cluster/k8s.js]: you can export other vc functions and throw error directly

File: src/rest-server/src/config/v2/hived.js (hunk 1 above)
Comment (mzmssg) [reply src/rest-server/src/config/v2/hived.js]: Configmap

File: src/rest-server/src/config/vc.js (hunk 2 above)
Comment (mzmssg) [reply src/rest-server/src/config/vc.js]: It's defined as k8s memory style of `Gi, Mi`, not MB

File: src/rest-server/src/models/v2/job/k8s.js (hunk 3 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/job/k8s.js]: yeah, it's a redundent check to avoid unnecessary logic dependency. Anyway, removal is also reasonable, will remove them

File: src/rest-server/src/config/v2/hived.js (hunk 8 above)
Comment (mzmssg) [reply src/rest-server/src/config/v2/hived.js]: both

File: src/rest-server/src/models/v2/virtual-cluster/index.js (hunk 6 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/virtual-cluster/index.js]: It's a workaround to avoid unnecessary copy, and we could overwrite yarn's implementation one by one

File: src/rest-server/src/models/v2/virtual-cluster/k8s.js (hunk 5 above)
Comment (mzmssg) [reply src/rest-server/src/models/v2/virtual-cluster/k8s.js]: It's doomed of a class, I think it's no need to unwrapper to functions.

File: src/rest-server/src/middlewares/v2/protocol.js (hunk 9 above)
Comment (mzmssg) [reply src/rest-server/src/middlewares/v2/protocol.js]: ok

File: src/rest-server/src/config/vc.js [hunk 11]
```diff
@@ -41,11 +42,24 @@ const vcStatusPutInputSchema = Joi.object().keys({
...
+        memory: '8192Mi',
+      },
+    };
+    logger.warn(`Hived enabled but spec not found or illegal: ${launcherConfig.hivedSpecPath}`);
+    logger.warn(`Init hived resource unit to: ${resourceUnits}`);
+  }
 } else {
   resourceUnits = {
-    DEFAULT: {
+    'null': {
```
Comment (abuccts) [inline src/rest-server/src/config/vc.js]: why changed to `'null'` here?

File: src/rest-server/src/middlewares/v2/protocol.js [hunk 12]
```diff
@@ -70,11 +73,16 @@ const convertPriority = (priorityClass) => {
 };
 
 const hivedValidate = (protocolObj) => {
+  if (!hivedSchema.validate(protocolObj)) {
```
Comment (abuccts) [inline src/rest-server/src/middlewares/v2/protocol.js]: could you add another hived middleware file and import here?

File: src/rest-server/src/config/vc.js (hunk 11 above)
Comment (mzmssg) [reply src/rest-server/src/config/vc.js]: It means the sku of `null` gpuType

File: src/rest-server/src/middlewares/v2/protocol.js (hunk 12 above)
Comment (mzmssg) [reply src/rest-server/src/middlewares/v2/protocol.js]: move hived validate to seprate file
